File: D1L1/MissingParentheses.cpp
#include <vector>
#include <string>
using namespace std;

int countCorrections(string const &par) {
    int p = 0, q = 0;
    for (int i = 0; i != par.size(); ++i) {
        if (par[i] == '(') {
            ++p;
        } else {
            if (p == 0) {
                ++q;
            } else {
                --p;
            }
        }
    }
    return p + q;
}

File: D1L1/MissingParentheses.hs
module MissingParentheses where 

countCorrections :: String -> Int
countCorrections par = f par 0 0 
    where f [] p q = p + q
          f ('(':xs) p q = f xs (p + 1) q
          f (')':xs) p q = if p > 0 then f xs (p - 1) q
                                    else f xs p (q + 1)
